Edwards25519

man2706kum opened the pull request:
Review can be done for the following field arithmetic in the field Fp(p=2^255-19):
- fn add
- fn subtract
- fn multiply
- fn reduce
- fn scalar_mult
- all the sub-functions on which the above functions are dependent.

The element in the field is defined in base radix51.
Testing fn inverse (field_element.sw) and fn double (ge25519.sw) are not possible as of now because of multiple calls to multiply and square fn.
Running multiple tests simultaneously for fn multiply is not possible as of now.

ewynx committed:
1: Addition BigUint

- Added datatype BigUint: vector of u32 limbs
- Implements addition for BigUint
- Tests for addition in main.sw (this is a script)

ewynx committed:
2: Subtraction BigUint

- Added subtraction for biguint
- With tests
- Introduced trait Zero, so we can check for "zero-being"
- Added tests for addition and structured them

ewynx committed:
Removed old TODO

ewynx committed:
Fix PR comment: make sure the edge case of 2 elements of length 1 for subtraction is covered.
- i >= 0 instead of i > 0
- test added for 2 elements of len 1
- added code for edge case 0 - 0 = 0
- test added for 0 - 0 = 0

ewynx committed:
Prep for 3: Karatsuba multiplication.
- Added the schoolbook mult from the swayPractice repo that works for 2 tests (not extensively tested though)
- Added data that can be used for a test case, with the values that would be used in the first iteration of Karatsuba

ewynx committed:
Larger test works now, which can be used for a small Karatsuba example.

ewynx committed:
WIP Karatsuba just 1 level deep
- A function to try out how Karatsuba would work splitting 1 time. Assumes a perfect scenario where x and y have the same length which is a multiple of 2.
- Working on going through the code step by step to see if it is doing the expected thing, according to the test data above the testcase that is called "biguint_schoolbook_mult"

man2706kum committed:
mod operation

man2706kum committed:
mod operation

ewynx committed:
1-Level Karatsuba for vector of length 2n finished up. This is not doing any checks and some unsafe unwrapping.

ewynx committed:
Added some checks wrt working of the 1 level deep mult.
- Length of x should be a multiple of 2
- At 2 places an unwrap was being done, it is now asserted first that this should be possible

ewynx committed:
Merge branch 'modularArithmetic' of github.com:hashcloak/fuel-crypto into modularArithmetic
This takes into account the changes made to 3-karatsub-mult.

ewynx committed:
Updated comment

ewynx committed:
Worked on modular arithmetic.
- a very naive implementation with 2 tests. Where modular reduction is done, the result is wrong by 1 - can't figure out why yet.
- added a comment about the "zero" definition for BigUint which we should get straight

Note that I added an equals function in the big_uint and named the mod function biguint_mod2, this is mainly because the compiler wouldn't stop giving issues. I think 'forc clean' doesn't actually clean everything :(

ewynx committed:
This starts the arithmetic functionality finite field with p = 2^255-19 and elements represented radix 51.
- Radix-51 representation
- Zero element
- Carry_propagate function
- Corresponding tests and test_helpers functions

ewynx committed:
Added mod_25519 with tests. Also improved the tests for carry_propagate.

ewynx committed:
WIP addition mod p.
The a+a test reveals that reduction is probably not correct - should be cheched.

man2706kum committed:
changes in mod_25519

man2706kum committed:
subtraction and multiplication

ewynx committed:
- Fixed the mod_25519 function (it was not reduces enough rounds in some situations)
- Added extra test for mod that represents this situation
- Both addition tests are now working

ewynx committed:
Merge branch 'edwards25519' of github.com:hashcloak/fuel-crypto into edwards25519

ewynx committed:
Addition to merge

ewynx committed:
Scalar multiplication with a scalar with max 32 bits.
- Added scalar mult function
- With corresponding tests

ewynx committed:
- Renamed mod_25519 to reduced, since it follows the 'reduced' function implementation from the go implementation
- Made the implementation constant (following the ref impl)
- Replaced the * 19 by the function times19
- Replaced the mod_25519 call in the addition function with the propagate_carry call
- Fixed the addition test that was using a test element too big

ewynx committed:
Added constant time equals function with tests.

ewynx committed:
Start on curve arithmetic

man2706kum committed:
multiply64 changed, a*19 replaced by times19(a), subtraction & multiply64 test added

man2706kum committed:
overflow for fn multiply fixed

ewynx committed:
- Fixed multiplication of 2 field elements by fixing add64 and add_multiply_64
- Replaced *19 with times19 from the NaCl impl
- Added tests and organized tests into which ones work if all ran at the same time and the mult test that...:
- Mult tests only work when ran 1 at a time -> open issue

man2706kum committed:
fn inverse added, tests for fn square added

ewynx committed:
Run the multiply test

ewynx committed:
Testing field_element library through contract

ewynx committed:
Making sure there is no recursion happening by accident in the contract

ewynx committed:
- Tests go through script again for now, since contract testing doesn't support much
- Split up tests in different files
- Some code and test cleanup + ran formatter as far as possible

ewynx committed:
Added a test for the inverse function (which hasn't been tested yet). But it doesn't terminate.
This is an example of what we need to be able to run, but currently doesn't seem possible.

ewynx committed:
Added a test for the inverse function (which hasn't been tested yet). But it doesn't terminate.
This is an example of what we need to be able to run, but currently doesn't seem possible.

ewynx committed:
Merge branch 'edwards25519' of github.com:hashcloak/fuel-crypto into edwards25519

ewynx committed:
Fixes for most of the warnings

ewynx committed:
Cleanup & TODO added

man2706kum committed:
few comments added

man2706kum committed:
bignum-lib removed

Mikerah commented on edwards25519/Forc.toml:
```diff
@@ -0,0 +1,7 @@
+[project]
+authors = ["HashCloak"]
+entry = "main.sw"
```
Since we are now running the tests in Rust, this should instead point to `field_element.sw`

Mikerah reviewed (changes requested):
I would also add that we should modify the README to include instructions on how to run everything. Not only does this help with PR reviews but someone looking at the repo can easily attempt to at least run the tests if they want and then try to integrate the library into their own project.

ewynx committed:
Multiply and square tests through contracts. First steps.

ewynx committed:
Moved subtraction tests

ewynx committed:
Moved addition tests

ewynx committed:
Converted reduction tests

ewynx committed:
Converted helpers64 and scalar_mult to testing with Rust.

ewynx committed:
Converted rshift tests to Rust

ewynx committed:
Updated README

ewynx committed:
Additions README

ewynx committed:
- Changed a line in the inverse implementation (field_element.sw)
- Added a test for inverse
- Running the inverse function partially gives an error: this is commented in the code

ewynx committed:
Running a few steps of inverse gives error.
